### PR TITLE
Added `oids_of!` macro

### PR DESCRIPTION
**Pull request by YohDeadfall**
That's an improvement for SPI making it less cumbersome. Just to feel the difference look at the tests. Who would really want to write something like `PgBuiltInOids::INT4OID.oid()` instead of just telling a Rust type?

I will change the behavior of this macro a bit when you will be ready for shipping `0.13`, so it won't produce any vector but a slice.

### CODE REVIEW

File: pgrx/src/datum/mod.rs [hunk 1]
```diff
@@ -149,3 +149,20 @@ impl<'src> Datum<'src> {
...
+/// ```
+/// use pgrx::{oids_of, datum::IntoDatum};
+/// let v = oids_of![i32, f64];
+/// assert_eq!(v[0], i32::type_oid().into());
+/// assert_eq!(v[1], f64::type_oid().into());
+/// ```
+#[macro_export]
+macro_rules! oids_of {
+    ($($t:path),* $(,)?) => ({
+        vec![$(::pgrx::pg_sys::PgOid::from(<$t>::type_oid())),*]
```
Comment (workingjubilee) [inline pgrx/src/datum/mod.rs]: I don't think this should include a `vec![]` when it could instead be invoked as `vec![oids_of!(...)]`

this would allow constructing arrays also, or other structures.

Comment (YohDeadfall) [reply pgrx/src/datum/mod.rs]: Yes, that's the plan when I get rid of `Option<Vec<PgOid>>` because now it's too wordy.

Comment (workingjubilee) [reply pgrx/src/datum/mod.rs]: ...if that's the plan, then?

Comment (YohDeadfall) [reply pgrx/src/datum/mod.rs]: Yeah, I will remove `vec!` from here and will change the methods of `SpiClient` to use slices as soon as you that the breaking change time has come (:

Comment (workingjubilee) [reply pgrx/src/datum/mod.rs]: I'd rather not add this then just break it again.

Comment (YohDeadfall) [reply pgrx/src/datum/mod.rs]: Let it be non-breaking then.

**Review by workingjubilee — CHANGES_REQUESTED** [review]
I don't think this has the right semantics right now. It's not clear to me that `oids_of!()` should be an `InvalidOid`, instead of, say, a compile-time error. Or an empty array.

**Comment by workingjubilee** [comment]
I think we should ship a non-plural `oid_of!` (I'd prefer an fn if we can swing it, tbh, but I'll settle) first and tinker with it a bit before jumping to generating arrays.

**Comment by YohDeadfall** [comment]
If something isn't an `IntoDatum` then a compiler error will appear, so the value of each element purely depends on the implementation of each type used there. `InvalidOid` is just to make an empty version of `oids_of!` to be strongly typed and not to depend on the context.

Let's say, I don't want `oids_of!` to give you an empty array of integers like in the following case:

```rust
let _: &[i32] = oids_of![];
```

That's why there's `[$crate::pg_sys::PgOid::Invalid; 0]`.

**Comment by YohDeadfall** [comment]
> I think we should ship a non-plural `oid_of!`

You have it already, `IntoDatum::type_oid()`, and it's a function. I thought about making a non-plural version of the macro, but I see no real use case for it.

**Comment by workingjubilee** [comment]
tbh, that's kind of the problem.

A macro creates a false impression this will resolve statically.

That fn produces runtime lookups.

**Comment by YohDeadfall** [comment]
It depends, actually. For built-in types, it's static, but for custom types it isn't.

**Comment by workingjubilee** [comment]
I realize that for the existing inputs, it's somewhat incoherent to use. But the reason to make it a zero-sized array, effectively, is to make a macro-call to it... possibly itself from a macro-expansion... valid anywhere we ask for `&[pg_sys::Oid]`, without having to reason about the exact input. For a given input, sure, they might need to do something like
```rust
let oids = oids_of![$($types,)];
oids.is_empty().then_some(&oids)
```
but this enables that.

Basically, we shouldn't overfit it to the existing use-cases.

**Comment by YohDeadfall** [comment]
Sorry, not following. Could you reword it a bit so I can get what I should do?

**Comment by workingjubilee** [comment]
I would still prefer zero arguments to evaluate to a zero-sized slice, or to [`compile_error`](https://doc.rust-lang.org/std/macro.compile_error.html)

**Comment by YohDeadfall** [comment]
Isn't `[$crate::pg_sys::PgOid::Invalid; 0]` a zero sized array?

**Comment by workingjubilee** [comment]
...hm.

I somehow managed to misread this code repeatedly, I suppose.

File: pgrx/src/datum/mod.rs [hunk 2]
```diff
@@ -149,3 +149,23 @@ impl<'src> Datum<'src> {
 /// A tagging trait to indicate a user type is also meant to be used by Postgres
 /// Implemented automatically by `#[derive(PostgresType)]`
 pub trait PostgresType {}
+
+/// Creates a [`Vec<pg_sys::Oid>`] containing identifiers of the provded types.
```
Comment (workingjubilee) [inline pgrx/src/datum/mod.rs]:
```suggestion
/// Creates an array of [`pg_sys::Oid`] with the OID of each provided type
```

File: pgrx/src/datum/mod.rs [hunk 3]
```diff
@@ -149,3 +149,23 @@ impl<'src> Datum<'src> {
...
+///
+/// # Examples
+///
+/// ```
+/// use pgrx::{oids_of, datum::IntoDatum};
```
Comment (workingjubilee) [inline pgrx/src/datum/mod.rs:213]:
```suggestion
/// use pgrx::{oids_of, datum::IntoDatum};
///
```

File: pgrx/src/datum/mod.rs [hunk 4]
```diff
@@ -149,3 +149,23 @@ impl<'src> Datum<'src> {
...
+/// let v = oids_of![i32, f64];
+/// assert_eq!(v[0], i32::type_oid().into());
+/// assert_eq!(v[1], f64::type_oid().into());
+/// ```
```
Comment (workingjubilee) [inline pgrx/src/datum/mod.rs:223]: mostly for the last line, because I'm silly
```suggestion
///
/// // the usual conversions or coercions are available
/// let oid_vec: Vec<pg_sys::Oid> = oids_of![i8, i16].to_vec();
/// let no_oid: &[pg_sys::Oid] = &oids_of![];
/// assert_eq!(no_oid.len(), 0);
/// ```
```

File: pgrx/src/datum/mod.rs [hunk 5]
```diff
@@ -149,3 +149,23 @@ impl<'src> Datum<'src> {
...
+#[macro_export]
+macro_rules! oids_of {
+    () =>(
+        [$crate::pg_sys::PgOid::Invalid; 0]
```
Comment (workingjubilee) [inline pgrx/src/datum/mod.rs:228]:
```suggestion
        // avoid coercions to an ambiguously-typed array or slice
        [$crate::pg_sys::PgOid::Invalid; 0]
```

**Review by workingjubilee — APPROVED** [review]
Sorry, I managed to not give this proper attention the past times I glanced at it. Some nits and a rebase should make this good.

File: pgrx/src/datum/mod.rs (hunk 4 above)
Comment (YohDeadfall) [reply pgrx/src/datum/mod.rs:223]: Not sure about it, so applied it partly. I made it mostly to demonstrate that the resulting type of `oids_of!` is known even for an empty set of types.

**Review by workingjubilee — APPROVED** [review]
